app/javascript/controllers/time_series_chart_controller.js

```diff
@@ -0,0 +1,424 @@
+import { Controller } from "@hotwired/stimulus"
+import tailwindColors from "@maybe/tailwindcolors"
+import * as d3 from "d3"
+
+const CHART_TYPES = [ "mini", "full" ]
+
+export default class extends Controller {
+  static values = { series: Object, chartType: String }
+
+  #_d3MainSvg = null
+  #_d3FullChartGroup = null
```

Comment:
Underscored because these are memoized further below, whereas the other properties aren't.